Add Sourcify FAQ post to the blog (#79)
ethereum/solidity-blog: _posts/2020-06-24-sourcify-faq.md

## Source Verification

### What is source verification?

In a software development context, the term source verification describes the process of verifying the legitimacy of code 
before interacting with it. Solidity contract source code can be verified by checking whether the on-chain deployed code 
is matching with published source code files. In order to allow for source verification, 
the source and metadata files need to be published, thus enabling developers to verify that the deployed bytecode is an exact match to the 
published files. By doing so, it can be verified that the contract code is exactly the same as when it was deployed on the
blockchain and has not been adjusted or tampered with before deploying it.

### Why is source verification important?

Without open source contracts it is hard to tell what the application you are using will actually be doing. 
One of the decentralized web's biggest aspriations is to build a trustless infrastructure, aiming to reduce the need to 
trust third parties and service providers and decreasing censorship opportunities. It should, by design, not be required to 

> "blindly" trust a dapp developer or web3 project. It shall, in the best case, always be possible to verify it for yourself, 

Author: From our perspective always, but there's also still closed source applications out there. Not sure if we want to be so strict?

> "blindly" trust a dapp developer or web3 project. It shall, in the best case, always be possible to verify it for yourself, 

Reviewer: I mean, I understand the meaning, but the expression caught my eye because it seems a little contradictory :) I'd say that a closed source application requires some trust by design so the previous sentence already makes a firm statement. If we're saying that, I see no reason not to say "always" here.

> "blindly" trust a dapp developer or web3 project. It shall, in the best case, always be possible to verify it for yourself, 

Author: Right, makes sense.

Sourcify adds the option to make any contract idenpendently verifyable. It also covers another important aspect: 
the availability of the code. Should EtherScan ever go down, there's no way to access their verified contracts anymore. 
Sourcify tries to solve this by making the storage of the source files and the processes around the verfication more 
decentralized, independent and hence more resilient.

### How is source verification & commeting your code with NatSpec connected and why is it more powerful if you do both?

If you verified that the code you deployed is an exact match to what you open-sourced that is great! 
Developers now can verify themselves that your dapp is actually doing what it promised.

However, it will still be hard for end-users to understand what they are doing when interacting with your code. 
Only by translating what the contract interactions mean into NatSpec you can help them comprehend what's going on under the 
hood. To ensure that you don't comment code with inadequate comments to mislead the users, all of the comments and metadata 

> also gets published and verified.

Author: > Do you mean that comments and metadata are published along with the bytecode so people who understand both can easily compare them and see if they don't match?

Yes. If you try to "cheat" it's basically out there in the open.

> also gets published and verified.

Reviewer 2: Maybe to clarify: You can of course always write wrong natspec comments in your code. What we guarantee is that the comments are the same that were used at compilation time (unless the deployer changed the metadata hash - but then they would also be able to change the comments themselves to begin with).

> also gets published and verified.

Reviewer: Makes sense. Maybe it could be said more explicitly?

> also gets published and verified.

Author: Tried to make this more explicit now. :)

You probably already stumbled over a transaction which had random hex string in the description. If you went ahead and 
signed this transaction you blindly trusted the developer of the dapp to do what they promised. 

Wouldn't it be great to have the option to verify what is happening and take an informed decision instead?  

Sourcify aims to add that needed context to signing decisions by displaying whether the source code has been verified and what 
the contract interaction actually means in layman's terms.

### How can we improve the UX using Sourcify?

Developers can help users take an informed decision by commenting their code using NatSpec. To verify that their NatSpec 
comments are actually accurate, they should publish the source files to an open directory.

It's important that interface tools like wallets display the NatSpec comments to make them visible to the user before 
the contract interaction.

### How can I check whether a contract has been verified?

Currently, contracts that have been verified via Sourcify don't yet get a special label or tag on wallets or interfaces 
(like blockexplorers). It would be great it that were the case in future though! 

> For now, you can use the Sourcify contract repository. It can be accessed via [https](https://contractrepo.komputing.org/) 

Reviewer: ```suggestion
For now, you can use the Sourcify contract repository. It can be accessed via [HTTPS](https://contractrepo.komputing.org/) 
```

or [IPFS](https://gateway.ipfs.io/ipns/QmNmBr4tiXtwTrHKjyppUyAhW1FQZMJTdnUrksA9hapS4u).

> Use the contract's address and search for it in the repository. If is is there, it is a match (aka verified). 🎉

Reviewer: ```suggestion
Use the contract's address and search for it in the repository. If it is there, it is a match (aka verified). 🎉
```

### What are full matches?

> Full matches mean that both the source files as well as the meta data files were an exact match between the deployed bytecode 

Reviewer: ```suggestion
Full matches mean that the source files as well as the metadata files were an exact match between the deployed bytecode 
```

and the published files.

### What are partial matches?

Partial matches occur when the metadata hash does not match. This means that the source code compiles to the same bytecode 
and thus the contract behaves in the same way, but the source code can be different: Variables can have misleading names, 
comments can be different and especially the NatSpec comments could have been modified.

## The Tooling Side

### What tooling is needed to make source verification more convenient?

The right support by existing and new tools can help make source verification much easier and more convenient for developers.

Quick wins on the tooling side include:
- Encourage auto-publishing of source files via deployment tools and make publishing the default option if license is 
specified as open source (licence identifiers are supported as of [Solidity v0.6.8](https://github.com/ethereum/solidity/releases/tag/v0.6.8)).
- Create libraries (in different languages) that wallets can use to easily recompile bytecode.

### What is needed on the wallet / UI side to make users aware of verified contracts?

> Walltes and user interfaces are a crucial puzzle piece to bring awareness and broader adoption to source verification and 

Reviewer: ```suggestion
Wallets and user interfaces are a crucial puzzle piece to bring awareness and broader adoption to source verification and 
```

NatSpec comments. To make users aware of verified contracts, wallets and other user interfaces could:

- Show which contracts are verified and which are not during contract interaction.
- Warn users if a contract has not been verified.
- Always display the NatSpec comments, if available.
- Warn users if NatSpec comments have not been verified and educate about the fact, that those explanations could potentially lie.

> - Educate users to only sign an interaction which is not NatSpec commented if they know what they are doing.

Reviewer: ```suggestion
- Educate users to only sign an interaction which is not NatSpec-commented if they know what they are doing.
```
